### PR TITLE
ported tg improved buildmode

### PR DESCRIPTION
**Pull request by harryob**
## About The Pull Request

admin bus is now 10x as effective (in build mode)

https://cdn.discordapp.com/attachments/539898120431992862/1030974347046822028/CleanShot_2022-10-15_at_14.48.48.mp4 https://cdn.discordapp.com/attachments/539898120431992862/1030989280539066459/CleanShot_2022-10-16_at_00.40.04.mp4

## Why It's Good For The Game

the old build mode was horribly coded on the backend and was also ugly. this one is better (from tg with various authors throughout the years)

## Changelog

:cl:
admin: build mode 2 - a bunch of new buttons (and better existing buttons)
/:cl:

### CODE REVIEW

**Comment by harryob** [comment]
idb2 fail making me look like a shitcoder

**Comment by harryob** [comment]
CleanShot 2022-10-16 at 03 32 56@2x
huh

**Comment by carlarctg** [comment]
pretty suspicious to add in build mode when you wont even benefit from it as a nondev.....

**Comment by harryob** [comment]
> pretty suspicious to add in build mode when you wont even benefit from it as a nondev.....

wait till he finds out the Wiki Maintainer is an admin

**Review by nauticall — APPROVED** [review]
sprites are fine! it's just admin tools anyway

File: code/modules/buildmode/submodes/variable_edit.dm [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	selected_key = null
+	selected_value = null
+
+/datum/buildmode_mode/varedit/change_settings(client/c)
+	var/list/locked = list("vars", "key", "ckey", "client", "icon")
+
+	selected_key = input(usr,"Enter variable name:" ,"Name", "name")
+	if(selected_key in locked && !check_rights(R_DEBUG,0))
+		return TRUE
+	var/type = tgui_input_list(usr,"Select variable type:" ,"Type", list("text","number","mob-reference","obj-reference","turf-reference"))
```
Comment (stanalbatross) [inline code/modules/buildmode/submodes/variable_edit.dm:26]: definesss

File: code/modules/buildmode/submodes/delete.dm [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+							qdel(Obj)
+						CHECK_TICK
+					if(!i)
+						to_chat(usr, "No instances of this type exist")
+						return
+					log_admin("[key_name(usr)] deleted all instances of type [O_type] ([i] instances deleted) ")
+					message_admins(SPAN_NOTICE("[key_name(usr)] deleted all instances of type [O_type] ([i] instances deleted) "))
+				if("Type and subtypes")
+					var/i = 0
+					for(var/Obj in world)
```
Comment (stanalbatross) [inline code/modules/buildmode/submodes/delete.dm:47]: no caps in var names

File: code/modules/buildmode/submodes/delete.dm [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+				return
+
+			if(tgui_alert(usr,"Second confirmation required. Delete?",,list("Yes","No")) != "Yes")
+				return
+
+			var/O_type = deleting.type
+			switch(action_type)
+				if("Strict type")
+					var/i = 0
+					for(var/atom/Obj in world)
```
Comment (stanalbatross) [inline code/modules/buildmode/submodes/delete.dm:35]: same

File: code/modules/buildmode/submodes/delete.dm [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+				return
+			var/atom/deleting = object
+			var/action_type = tgui_alert(usr,"Strict type ([deleting.type]) or type and all subtypes?",,list("Strict type","Type and subtypes","Cancel"))
+			if(action_type == "Cancel" || !action_type)
+				return
+
+			if(tgui_alert(usr,"Are you really sure you want to delete all instances of type [deleting.type]?",,list("Yes","No")) != "Yes")
+				return
+
+			if(tgui_alert(usr,"Second confirmation required. Delete?",,list("Yes","No")) != "Yes")
```
Comment (stanalbatross) [inline code/modules/buildmode/submodes/delete.dm:28]: null args 💢

File: code/modules/buildmode/submodes/delete.dm [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+			T.ScrapeAway()
+		else if(isatom(object))
+			qdel(object)
+
+	if(LAZYACCESS(modifiers, RIGHT_CLICK))
+		if(check_rights(R_DEBUG|R_SERVER)) //Prevents buildmoded non-admins from breaking everything.
+			if(isturf(object))
+				return
+			var/atom/deleting = object
+			var/action_type = tgui_alert(usr,"Strict type ([deleting.type]) or type and all subtypes?",,list("Strict type","Type and subtypes","Cancel"))
```
Comment (stanalbatross) [inline code/modules/buildmode/submodes/delete.dm:21]: null args

File: code/modules/buildmode/submodes/boom.dm [hunk 6]
```diff
@@ -0,0 +1,26 @@
+/datum/buildmode_mode/boom
+	key = "boom"
+	help = "Click on Mode = Set explosive settings\n\
+	Mouse Button on obj = Kaboom\n\
+	NOTE: Using the \"Launch Supplypod\" verb allows you to do this in an IC way (i.e., making a cruise missile come down from the sky and explode wherever you click!)"
```
Comment (stanalbatross) [inline code/modules/buildmode/submodes/boom.dm:5]: isn't the actual name of the verb different

File: code/modules/buildmode/buttons.dm [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+/atom/movable/screen/buildmode/modeswitch/New(bld, mt)
+	modetype = mt
+	icon_state = "buildmode_[initial(modetype.key)]"
+	name = initial(modetype.key)
+	return ..(bld)
+
+/atom/movable/screen/buildmode/modeswitch/clicked()
+	bd.change_mode(modetype)
+	return 1
```
Comment (stanalbatross) [inline code/modules/buildmode/buttons.dm:66]: TRUE or defines, no magic numbers

File: code/modules/buildmode/buttons.dm [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	screen_loc = "NORTH,WEST+2"
+	name = "Change Dir"
+
+/atom/movable/screen/buildmode/bdir/update_icon()
+	dir = bd.build_dir
+
+/atom/movable/screen/buildmode/bdir/clicked()
+	bd.toggle_dirswitch()
+	update_icon()
+	return 1
```
Comment (stanalbatross) [inline code/modules/buildmode/buttons.dm:52]: same

File: code/modules/buildmode/buttons.dm [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+	icon_state = bd.mode.get_button_iconstate()
+
+/atom/movable/screen/buildmode/help
+	icon_state = "buildhelp"
+	screen_loc = "NORTH,WEST+1"
+	name = "Buildmode Help"
+
+/atom/movable/screen/buildmode/help/clicked()
+	bd.mode.show_help(usr.client)
+	return 1
```
Comment (stanalbatross) [inline code/modules/buildmode/buttons.dm:39]: same

File: code/modules/buildmode/buttons.dm [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+	screen_loc = "NORTH,WEST"
+
+/atom/movable/screen/buildmode/mode/clicked(location, var/list/modifiers)
+	if(LAZYACCESS(modifiers, LEFT_CLICK))
+		bd.toggle_modeswitch()
+	else if(LAZYACCESS(modifiers, RIGHT_CLICK))
+		bd.mode.change_settings(usr.client)
+
+	update_icon()
+	return 1
```
Comment (stanalbatross) [inline code/modules/buildmode/buttons.dm:27]: same

File: code/modules/buildmode/buildmode.dm [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+	close_switchstates()
+	mode = new newmode(src)
+	mode.enter_mode(src)
+	modebutton.update_icon()
+
+/datum/buildmode/proc/change_dir(newdir)
+	build_dir = newdir
+	close_dirswitch()
+	dirbutton.update_icon()
+	return 1
```
Comment (stanalbatross) [inline code/modules/buildmode/buildmode.dm:139]: 💢

**Review by stanalbatross — CHANGES_REQUESTED** [review]

**Review by stanalbatross — CHANGES_REQUESTED** [review]
BUILD BASIC
- make lmb + ctrl place a framed r-window not directional
- dir changing does not work with airlocks
- downgrading rwalls just deletes them
- you cannot upgrade rwalls to heavy rwalls
KABOOM
- help tip needs to be "right click on mode"
FILL MODE
- should be able to pick object type via right clicking ic object like with build advanced

**Review by stanalbatross — APPROVED** [review]

**Pull request merged**